src/ComServerHelpers/ComServer.cs

```diff
@@ -0,0 +1,239 @@
+{
+    /// <summary>
+    /// Map of class factories and the registration cookie from the CLSID that the factory creates.
+    /// </summary>
+    private readonly Dictionary<Guid, (BaseClassFactory factory, uint cookie)> factories = [];
+
+    private readonly StrategyBasedComWrappers comWrappers = new();
+
+    /// <summary>
+    /// Tracks the creation of the first instance after server is started.
+    /// </summary>
+    private TaskCompletionSource<object>? firstInstanceCreated;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ComServer"/> class.
+    /// </summary>
+    public unsafe ComServer()
+    {
+        using ComPtr<IGlobalOptions> options = default;
+        Guid clsid = CLSID_GlobalOptions;
+        Guid iid = IGlobalOptions.IID_Guid;
+        if (CoCreateInstance(&clsid, null, CLSCTX.CLSCTX_INPROC_SERVER, &iid, (void**)options.GetAddressOf()) == HRESULT.S_OK)
+        {
+            options.Get()->Set(GLOBALOPT_PROPERTIES.COMGLB_RO_SETTINGS, (nuint)GLOBALOPT_RO_FLAGS.COMGLB_FAST_RUNDOWN);
+        }
+    }
+
+    /// <summary>
+    /// Register a class factory with the server.
+    /// </summary>
+    /// <param name="factory">The class factory to register.</param>
+    /// <param name="comWrappers">The implementation of <see cref="ComWrappers"/> to use for wrapping.</param>
+    /// <returns><see langword="true"/> if <paramref name="factory"/> was registered; otherwise, <see langword="false"/>.</returns>
+    /// <remarks>Only one factory can be registered for a CLSID.</remarks>
+    /// <exception cref="ObjectDisposedException">The instance is disposed.</exception>
+    /// <exception cref="ArgumentNullException"><paramref name="factory"/> or <paramref name="comWrappers"/> is <see langword="null"/>.</exception>
+    /// <exception cref="InvalidOperationException">The server is running.</exception>
+    /// <seealso cref="UnregisterClassFactory(Guid)"/>
+    public unsafe bool RegisterClassFactory(BaseClassFactory factory, ComWrappers comWrappers)
+    {
+        ObjectDisposedException.ThrowIf(IsDisposed, this);
+        ArgumentNullException.ThrowIfNull(factory);
+        ArgumentNullException.ThrowIfNull(comWrappers);
+        if (IsRunning)
+        {
+            throw new InvalidOperationException("Can only add class factories when server is not running.");
+        }
+
+        Guid clsid = factory.Clsid;
+
+        if (factories.ContainsKey(clsid))
+        {
+            return false;
+        }
+
+        factory.InstanceCreated += Factory_InstanceCreated;
+
+        nint wrapper = this.comWrappers.GetOrCreateComInterfaceForObject(new BaseClassFactoryWrapper(factory, comWrappers), CreateComInterfaceFlags.None);
+
+        uint cookie;
+        CoRegisterClassObject(&clsid, (IUnknown*)wrapper, CLSCTX.CLSCTX_LOCAL_SERVER, (REGCLS.REGCLS_MULTIPLEUSE | REGCLS.REGCLS_SUSPENDED), &cookie).ThrowOnFailure();
+
+        factories.Add(clsid, (factory, cookie));
```

Comment:
`wrapper` will leak if either of these throw.

Reply:
added try/catch/rethrow + free